New mod element system

Goldorion opened the pull request:
After moving to plugins and rewriting variables and resources, it's time for mod elements. However, moving mod elements to plugins requires several major changes. So, this PR simply starts bychanging the mod element type creation. Read the following changelog for all major changes:

**Changelog:**
1. Move `ModElementType` enum to a class
2. Remove usage of `@SerializedName` (see line 4)
3. Move all mod element type parameters to a single `ModElementType` constructor
4. Use a single string (`registryName`) for everything (except L10N label keys inside GUIs)*
5. Use one image file for each mod element type icon**
6. Re-order some elements to have them in alphabetic order (e.g. `BaseType`). Not important, but easier to find something
7. MET variables are now in an inner class called `BuiltInTypes` of `ModElementType`. Variables get their MET value inside the method of `ModElementTypeLoader`.
8. Due to the new system, `TestWorkspaceDataProvider#getExampleFor()` now has to use if statements. I tried to keep the switch, but nothing I tested worked.

*With time, some mod elements got a new name, like ranged item and living entity. However, serialized names were still the original names. To simplify the code, I decided to use a single name for everything (serialized names, registry names, etc.). However, as breaking older workspaces is bad, I added a small converter that is called every time a workspace is opened. This small converter simply replaces the MET `_type`  (mod element's file) and the MET `type` (workspace file). This way, we are sure we are not breaking old workspaces. I tested with a workspace created in 2021.1 containing an advancement, an entity and a ranged item. The advancement tested unchanged ME. The living entity and the ranged item have a new registry name, so they tested if everything still worked fine. Everything has been converted properly. 

**The file name of the MET's icon is its registry name.

To review:
- elements.registry.ModElementType
- elements.registry.ModElementTypeLoader
- elements.GeneratableElement (Converter)
- generator.DefinitionsProvider?
- workspace.Workspace (Converter)
- workspace.elements.ModElement
- integration.TestWorskpaceDataProvider?
All other files are new MET icons and changes due to the new names, package, inner classes and code adaptation (due to enum -> class)

KlemenDEV commented:
Indeed do this in multiple stages with multiple PRs, otherwise, this will be unreviewable/unmergable

Goldorion commented:
If tests don't fail (due to a refactoring I did to change the Mob class for LivingEntity), I finished making everything I wanted to change for this first PR.

Defeatomizer commented:
>[Template Generator] Failed to generate template: json/dimension/dimension_overworld.json.ftl
    freemarker.core.InvalidReferenceException: The following has evaluated to null or missing:
    ==> ge.getModElement().getType().name  [in template "mcitems.ftl" at line 104, column 18]

There's a problem:
```freemarker
<#if ge.getModElement().getType().name() == "FLUID">
```
This line should actually look like this:
```freemarker
<#if ge.getModElement().getType().getRegistryName() == "fluid">
```

Applies to all "**.ftl**" occurrences of `ge.getModElement().getType().name()` (and maybe some other strings :/ ).

Goldorion commented:
> 
> 
> > [Template Generator] Failed to generate template: json/dimension/dimension_overworld.json.ftl
> > freemarker.core.InvalidReferenceException: The following has evaluated to null or missing:
> > ==> ge.getModElement().getType().name  [in template "mcitems.ftl" at line 104, column 18]
> 
> There's a problem:
> 
> ```
> <#if ge.getModElement().getType().name() == "FLUID">
> ```
> 
> This line should actually look like this:
> 
> ```
> <#if ge.getModElement().getType().getRegistryName() == "fluid">
> ```
> 
> Applies to all "**.ftl**" occurrences of `ge.getModElement().getType().name()` (and maybe some other strings :/ ).

I noticed the problem, but thank you for the solution. :)

KlemenDEV commented on src/main/java/net/mcreator/element/registry/ModElementRegistry.java:
```diff
@@ -0,0 +1,128 @@
+public class ModElementRegistry {
+
+	public static final List<ModElementType<?>> ELEMENTS = new ArrayList<>() {{
+		add(ADVANCEMENT = new ModElementType<>("achievement", 'h', BaseType.ACHIEVEMENT, RecipeType.NONE,
```
Afaik achievements are now basetype datapack, please check others too again so we don't have mistakes here

KlemenDEV commented on src/main/java/net/mcreator/ui/modgui/LivingEntityGUI.java:
```diff
@@ -876,7 +876,7 @@ private void regenerateAITasks() {
 			if (hasErrors)
 				return new AggregatedValidationResult.MULTIFAIL(compileNotesPanel.getCompileNotes().stream()
-						.map(compileNote -> "Mob AI builder: " + compileNote.getMessage())
+						.map(compileNote -> "LivingEntity AI builder: " + compileNote.getMessage())
```
This is user text, so change this to Living entity AI ...

KlemenDEV commented on src/main/java/net/mcreator/workspace/Workspace.java:
```diff
@@ -488,6 +490,8 @@ public static Workspace createWorkspace(File workspaceFile, WorkspaceSettings wo
 
 	public void reloadFromFS() {
 		String workspace_string = FileIO.readFileToString(fileManager.getWorkspaceFile());
+		workspace_string = workspace_string.replaceAll("\"type\": \"mob\",", "\"type\": \"livingentity\",")
```
You are checking this in deserializer so you don't need this here too (I think)

Goldorion replied:
The code here will change the code in the .mcreator while the code in GeneratableElement will change it for the element files (worspaceFolder/elements).

KlemenDEV replied:
I see. But replacing like this is not ok, JSON can have different formatting. Check how adapter is registered for soundelements here: https://github.com/MCreator/MCreator/blob/064a81aa50bc3551951d93dcba9e877dc6481fa2/src/main/java/net/mcreator/workspace/elements/SoundElement.java#L89

Make a similar adapter for modelement to refactor the type cleanly with json parsing, not by replacing texts

KlemenDEV replied:
This was not resolved

Goldorion replied:
What's wrong in the commit? https://github.com/MCreator/MCreator/pull/1400/commits/347a7f391c24a67e92324e0411359a179b68f466

KlemenDEV commented on src/main/java/net/mcreator/element/registry/ModElementRegistry.java:
```diff
@@ -0,0 +1,128 @@
+
+import static net.mcreator.element.registry.ModElementRegistry.BuiltInTypes.*;
+
+public class ModElementRegistry {
```
Refactor to ModElementTypeRegistry, although it is a bit long, it describer the class more accurately

KlemenDEV replied:
Or maybe even better, keep this ModElementType and existing package, so eg. ModElementType.PROCEDURE like calls don't need refactoring. This will also reduce number of changed files a lot and make this reviewable. We can still refactor this later if it turns out it is needed

KlemenDEV commented on src/main/java/net/mcreator/element/registry/ModElementRegistry.java:
```diff
@@ -0,0 +1,128 @@
+	public static final List<ModElementType<?>> ELEMENTS = new ArrayList<>() {{
+		add(ADVANCEMENT = new ModElementType<>("achievement", 'h', BaseType.ACHIEVEMENT, RecipeType.NONE,
+				AchievementGUI::new, Achievement.class));
+		add(ARMOR = new ModElementType<>("armor", 'a', BaseType.ARMOR, RecipeType.ARMOR, ArmorGUI::new, Armor.class));
```
Maybe make register method so this registry is shaped a bit more like "MC registries"

KlemenDEV commented on src/main/java/net/mcreator/element/GeneratableElement.java:
```diff
@@ -101,14 +103,25 @@ public void setLastModElement(ModElement lastModElement) {
+				newType = "livingentity";
+				break;
+			}
+			jsonElement.getAsJsonObject().addProperty("_type", newType);
```
I don't think it makes sense to add property here as it won't be used in deserialization anymore

Goldorion replied:
I’m not familiar with the generator system yet, so I thought it would be useful for unused mod elements (so mod elements that aren’t saved again by the user).

KlemenDEV replied:
You were thinking correct, but this is not saved if you set it here. No need to save though as your code here will fix this each time until user edits it to parentally "fix" it

KlemenDEV reviewed (commented):
1.15.2 definition rename is incorrect, please test your changes in both versions by hand too.

Goldorion commented:
> 1.15.2 definition rename is incorrect, please test your changes in both versions by hand too.

I always do a last round of tests later to see if I forgot things. Thank you for the report.

Defeatomizer commented:
OKay, thanks for curing Living Entity!

KlemenDEV commented:
There are two conflicting files

Goldorion commented:
> 
> 
> There are two conflicting files

I don't see any conflicts.

KlemenDEV commented:
When PR is in draft, GitHub does not show it, but you can always check with eg. GitHub console to see them

KlemenDEV commented:
![image](https://user-images.githubusercontent.com/16374228/124127425-fd462f80-da7b-11eb-954a-e80c13bf9a38.png)

Goldorion commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/16374228/124127425-fd462f80-da7b-11eb-954a-e80c13bf9a38.png)

Yes, I know. It's caused by #1329.

Defeatomizer commented on src/main/java/net/mcreator/element/RecipeType.java:
```diff
@@ -17,6 +17,44 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2021, Pylo, opensource contributors
```
The header comment is duplicated. It's not very important, but please remove the odd comment.

Goldorion replied:
Intellij has some problems with them for some time. Max had this problem too with one of his PR.

KlemenDEV commented:
Also in light theme, there is one icon that is not included, but not sure which one

Defeatomizer reviewed (changes requested):
>Also in light theme, there is one icon that is not included, but not sure which one

Empty icon is not present in the **`default_light`** theme.
Tab icon is called [**`item_group.png`**](https://github.com/Goldorion/MCreator/blob/new-mod-element-system/plugins/mcreator-themes/themes/default_light/images/mod_types/item_group.png) for this theme (instead of **`tab.png`**), is that needed?

KlemenDEV commented:
Probably wrong name, @Goldorion please test the PR a bit more so we can make sure all works fine.

Goldorion commented:
> 
> 
> > Also in light theme, there is one icon that is not included, but not sure which one
> 
> Tab icon is called [**`item_group.png`**](https://github.com/Goldorion/MCreator/blob/new-mod-element-system/plugins/mcreator-themes/themes/default_light/images/mod_types/item_group.png) for the _`default_light`_ theme (instead of **`tab.png`**), is that needed?

In the beginning, I also changed the mod element ID of tabs for `item_group`, but I changed it back for `tab`. However, I forgot to change the name for the icon file.

KlemenDEV commented:
![image](https://user-images.githubusercontent.com/16374228/125203925-0f458080-e27b-11eb-9c10-61170f0e4e29.png)

One light theme icon is still missing

KlemenDEV commented on src/main/java/net/mcreator/element/converter/ConverterRegistry.java:
```diff
@@ -48,9 +48,9 @@
 public class ConverterRegistry {
 
-	private static final Map<ModElementType, List<IConverter>> converters = new HashMap<>() {{
+	private static final Map<ModElementType<?>, List<IConverter>> converters = new HashMap<ModElementType<?>, List<IConverter>>() {{
```
Since J9, second diamond expression is not needed

KlemenDEV commented on src/main/java/net/mcreator/generator/DefinitionsProvider.java:
```diff
@@ -34,28 +35,30 @@
 	private static final Logger LOG = LogManager.getLogger("Definition Loader");
 
-	private final Map<ModElementType, Map<?, ?>> cache = new ConcurrentHashMap<>();
+	private final Map<String, Map<?, ?>> cache = new ConcurrentHashMap<>();
```
Key should be ModElementType, just add equals and hashId that uses name for the source to ModElementType

KlemenDEV commented on src/main/java/net/mcreator/workspace/Workspace.java:
```diff
@@ -404,6 +404,8 @@ public static Workspace readFromFS(File workspaceFile, @Nullable Window ui)
 			String workspace_string = FileIO.readFileToString(workspaceFile);
 			Workspace retval;
 			try {
+				workspace_string = workspace_string.replaceAll("\"type\": \"mob\",", "\"type\": \"livingentity\",")
```
You marked this solved two times already. This needs proper implementation with deserializer.. Don't just mark things as resolved, please

KlemenDEV replied:
Actually I see you added this. My question is changed to why you are still doing parsing here if you added proper JSON adapter for this?

Goldorion replied:
I surely forgot to remove it.

KlemenDEV reviewed (changes requested):
Some more things I noticed

KlemenDEV commented on src/main/java/net/mcreator/ui/MCreatorApplication.java:
```diff
@@ -142,6 +143,10 @@ private MCreatorApplication(List<String> launchArguments) {
 		EntityAnimationsLoader.init();
 
+		// register mod element types
+		// TODO: Move mod elements to the plugin system
```
Master code should not have todos, I suggest you run own todo board for todo notes

KlemenDEV commented on src/main/java/net/mcreator/element/ModElementTypeRegistry.java:
```diff
@@ -1,117 +1,130 @@
+
+	public static List<ModElementType<?>> REGISTRY = new ArrayList<>();
+
+	public static void registryModElements() {
```
Rename to loadModElements

Goldorion replied:
I named it registerModElements, so when we will add support for custom mod elements, we won't have to rename ot again.

KlemenDEV replied:
Even then, it will load them, and all methods for loading things start with load so far

KlemenDEV commented on src/main/java/net/mcreator/element/ModElementType.java:
```diff
@@ -18,104 +19,142 @@
-	@SerializedName("overlay") OVERLAY(BaseType.OVERLAY, RecipeType.NONE),
-
-	@SerializedName("procedure") PROCEDURE(BaseType.PROCEDURE, RecipeType.NONE),
+public class ModElementType<GE extends GeneratableElement> implements Comparable<ModElementType<?>> {
```
You should add equals and hashcode to this class

Defeatomizer commented:
Is there any reason this should be ready before **2021.2**?

KlemenDEV commented:
> Is there any reason this should be ready before **2021.2** is released?

More time to test. I plan two more snapshots if all goes to the plans.

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -166,8 +167,7 @@ public static void fillWorkspaceWithTestData(Workspace workspace) {
 			boolean emptyLists, int valueIndex) {
-		switch (modElement.getType()) {
-		case ACHIEVEMENT:
+		if (ModElementType.ADVANCEMENT.equals(modElement.getType())) {
```
This should remain switch

KlemenDEV replied:
Or not, nevermind

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -480,7 +477,8 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
 			return overlay;
-		case GUI:
+		} else if (ModElementType.GUI.equals(modElement.getType())) {
+			ArrayList<GUIComponent> components;
```
Put this back, no reason to split it like this.

Goldorion replied:
Split what? Yesterday, you said it was good.
![image](https://user-images.githubusercontent.com/38427877/125517790-69ec2500-c220-49ad-9059-d20b2d14a614.png)

KlemenDEV replied:
if is ok, but no need to split array list like you did

Goldorion replied:
Oh, I didn't notice.

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -496,13 +494,10 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
-				components
-						.add(new net.mcreator.element.parts.gui.Label("text3 <VAR:test>", 100, 150, "text3 <VAR:test>",
-								Color.black, new Procedure("condition4")));
+				components.add(new Label("text", 100, 150, "text", Color.red, new Procedure("condition2")));
```
Run code reformat so this breaks back due to line width limit

Goldorion replied:
I reformatted the code and nothing happened. The code is already properly formatted. I'm also not sure to understand what you want as the current line (in green) is properly formatted.

KlemenDEV replied:
Oh, it is due to indentation change as there is no longer switch :)

KlemenDEV reviewed (changes requested):
Also when I open any workspace with entities or ranged items, workspace only partially loads and then breaks. Please re-test this PR.

Goldorion commented:
@KlemenDEV I found how to fix the converter for the workspace file. However, I don't find the solution for the generatable elements. I'm still "new" to the generator system, so I don't yet exactly how it works, so do you know where the converter to fix the `_fv` field should go?  Currently, I fix it directly into the deserializer of GSOthe subclass GSONAdapter, but it's called after the error. I tried different things, but nothing works. Do you have a tip to solve this problem? It's the only one remaining problem to have a working converter.

KlemenDEV commented:
I will take a look when I get some time

KlemenDEV reviewed (commented):
For me, generatable elements work, @Goldorion are you sure there are still any problems?

Goldorion commented:
> 
> 
> For me, generatable elements work, @Goldorion are you sure there are still any problems?

When I tested with an older workspace, I got problems converting the mod element type in the mod element's file (so the `type` (sorry, I said `_fv`). They weren't changed (so converted).

Defeatomizer commented on src/main/java/net/mcreator/workspace/elements/ModElement.java:
```diff
@@ -324,4 +323,26 @@ public void setParentFolder(@Nullable FolderElement parent) {
+				throws JsonParseException {
+			JsonObject json = jsonElement.getAsJsonObject();
+
+			String newType = json.get("type").getAsString();
+			if (newType.equals("gun")) {
+				newType = "rangeditem";
+			} else if (newType.equals("mob")) {
+				newType = "livingentity";
+			}
+
+			json.addProperty("type", newType);
+
+			return gson.fromJson(json, ModElement.class);
```
If the errors are caused by using `addProperty()` here, I'd suggest adding a private method to the `ModElement` class that returns its type as a string (e. g. `getTypeString()`) and change deserializing code like this:
```suggestion
			ModElement retVal = gson.fromJson(json, ModElement.class);

			if (retVal.getTypeString().equals("gun")) {
				retVal.setType(ModElementType.RANGEDITEM);
			} else if (retVal.getTypeString().equals("mob")) {
				retVal.setType(ModElementType.LIVINGENTITY);
			}

			return retVal;
```

KlemenDEV replied:
Previous approach is fine.

Defeatomizer reviewed (changes requested):
> When I tested with an older workspace, I got problems converting the mod element type in the mod element's file (so the `type` (sorry, I said `_fv`). They weren't changed (so converted).

I hope my notes can help to resolve the issues.

KlemenDEV commented on src/main/java/net/mcreator/element/ModElementTypeLoader.java:
```diff
@@ -125,6 +125,7 @@ public static void loadModElements() {
 		}
-		return null;
+
+		throw new RuntimeException("Mod element type " + modElementName + " is not a registered type");
```
This is intentional, as this case should not happen

Goldorion replied:
But it causes build errors as nothing is returned, so how do we fix it?

Defeatomizer replied:
>how do we fix it?

Addons code:
```freemarker
  <#list w.getElementsOfType("BLOCK") as mod>
```
Mod element types definition:
```java
		ModElementType.BLOCK = register(
				new ModElementType<>("block", 'b', BaseType.BLOCK, RecipeType.BLOCK, BlockGUI::new, Block.class));
```
ME types names are still UPPERCASED within addons.
This really works as designed.

Goldorion replied:
Oh ok, I'll fix them. However, this is strange because before, tests didn't fail.

Defeatomizer replied:
This is not the only place in the addons, tests will keep failing.
